docs/usage.md

```diff
@@ -1,3 +1,721 @@
+async def command(ctx: tanjun.abc.Context) -> None:
+    ...
+```
+
+There's a collection of standard checks in [tanjun.checks][] which are all
+exported top level and work with all the command types, the only
+configuration most users will care about for these is `error_message` argument
+which lets you adjust the response these gives when they fail but the
+permission checks also need a required permission to be passed positionally.
+
+```py
+component = (
+    tanjun.Component()
+    .add_check(tanjun.GuildCheck())
+    .add_check(tanjun.AuthorPermissionCheck(hikari.Permissions.BAN_MEMBER))
+    .add_check(tanjun.OwnPermissionCheck(hikari.Permissions.BAN_MEMBER))
+)
+
+@component.with_check
+async def db_check(ctx: tanjun.abc.Context, db: alluka.Injected[Db]) -> None:
+    if (await db..get_user(ctx.author.id)).banned:
+        raise tanjun.CommandError("You are banned from using this bot")
+
+    raise False
+
+
+@tanjun.with_owner_check(follow_wrapped=True)
+@tanjun.as_message_command("name")
+@tanjun.as_slash_command("name", "description")
+async def owner_only_command(ctxL tanjun.abc.Context):
+    ...
+```
+
+Checks (both custom and standard) can be added to clients, components and
+commands using either the chainable `add_check` method or the decorator
+style `with_check` method with the standard checks providing `with_...`
+decorators which can be applied to commands to add the check and checks
+on a client, component or command group will be used for every child
+command.
+
+```py
+def check(ctx: tanjun.abc.Context) -> bool:
+    if ctx.author.discriminator % 2:
+        raise tanjun.CommandError("You are not one of the chosen ones")
+
+    return True
+```
+
+A custom check can be implemented by making a function with either the signature
+`def (tanjun.abc.Context, ...) -> bool` or `async def (tanjun.abc.Context, ...) -> bool`
+where [dependency injection][dependency-injection] is supported, returning
+`True` indicates that the check passed and returning `False` indicates that the
+check failed and the client should continue looking for a matching command. You
+will most likely want to raise [CommandError][tanjun.errors.CommandError] to
+end command execution with a response rather than carrying on with the command
+search.
+
+### Execution hooks
+
+Command hooks are callbacks which are called around command execution, these
+are contained within [Hooks][tanjun.hooks.Hooks] objects which may be added
+to a command, client or component using `set_hooks` where hooks on a client,
+component or command group will be callde for every child command.
+
+There are several different kinds of hooks which all support dependency
+injection and may be synchronous or asynchronous:
+
+```py
+hooks = tanjun.AnyHooks()
+
+@hooks.with_pre_execution  # hooks.add_pre_execution
+async def pre_execution_hook(ctx: tanjun.abc.Context) -> None:
+    ...
+```
+
+Pre-execution are called before the execution of a command (so after command
```

Comment:
Rather than brackets do "but after command matching has finished and all the checks have passed"